File: native/swift/swift-export-standalone-integration-tests/simple/testData/execution/inheritance/multi_level_swift_hierarchy/multi_level_swift_hierarchy.swift
import Inheritance
import Testing

@Test
func twoSwiftSubclassLevelsPreserveSuperChainThroughKotlin() throws {
    class SwiftMiddle: BaseClass {
        override func baseFunction() -> String { "swift-middle>" + super.baseFunction() }
    }
    class SwiftLeaf: SwiftMiddle {
        override func baseFunction() -> String { "swift-leaf>" + super.baseFunction() }
    }

    let value = SwiftLeaf()
    #expect(value.baseFunction() == "swift-leaf>swift-middle>base-class")
    #expect(callFunBaseClass(value: value) == "swift-leaf>swift-middle>base-class")
}

@Test
func severalSwiftLevelsCallKotlinSuperWhichCallsBackToSwift() throws {
    class SwiftMiddle: SuperReentryBase {
        override func callback() -> String { "swift-middle-callback" }
        override func operation() -> String { "swift-middle>" + super.operation() }
    }

    class SwiftLeaf: SwiftMiddle {
        override func callback() -> String { "swift-leaf-callback" }
        override func operation() -> String { "swift-leaf>" + super.operation() }
    }

    let value = SwiftLeaf()
    #expect(value.operation() == "swift-leaf>swift-middle>kotlin-operation>swift-leaf-callback")
    #expect(callSuperReentry(value: value) == "swift-leaf>swift-middle>kotlin-operation>swift-leaf-callback")
}

File: native/swift/swift-export-standalone-integration-tests/simple/testData/execution/inheritance/multi_level_swift_hierarchy/multi_level_swift_hierarchy.kt
// KIND: STANDALONE
// FREE_COMPILER_ARGS: -opt-in=kotlin.native.internal.InternalForKotlinNative
// MODULE: Inheritance
// FILE: multi_level_swift_hierarchy.kt

open class BaseClass {
    open fun baseFunction(): String = "base-class"
}

fun callFunBaseClass(value: BaseClass): String = value.baseFunction()

open class SuperReentryBase {
    open fun callback(): String = "kotlin-callback"
    open fun operation(): String = "kotlin-operation>${callback()}"
}

fun callSuperReentry(value: SuperReentryBase): String = value.operation()
